# Wrangling and Analyzing Data from We Rate Dogs
## @rate_dogs on Twitter

## Gathering the Data ##

In [1]:
import pandas as pd
import numpy as np
import requests
import os
import tweepy
import json

In [2]:
# Establishing the first table: archive

archive = pd.read_csv('twitter-archive-enhanced.csv')

In [3]:
# Downloading image-predictions.tsv from Udacity's server

folder_name = "C:/Users/chris/OneDrive/Documents/Coding/Udacity_Data_Analyst/Wrangle_Data_Project"
file_name = 'image-predictions.tsv'
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'

response = requests.get(url)
with open(os.path.join(folder_name, file_name), mode='wb') as file:
	file.write(response.content)

In [4]:
# Establishing the second table: image_predictions

image_predictions = pd.read_csv('image-predictions.tsv', sep='\t')

In [5]:
# Setting up the Twitter API

consumer_key = 'REMOVED'
consumer_secret = 'REMOVED'
access_token = 'REMOVED'
access_secret = 'REMOVED'

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth)

In [6]:
# This variable is just to generate a percentage complete while the below loop runs since it takes a while.
number_of_tweets = archive.tweet_id.count()

tweet_json_file = 'tweet_json.txt'
current_record = 0

# Querying Tweepy to pull all available information for the tweets listed in the archive Udacity provided
for tweet_id in archive.tweet_id:
    try:
        tweet = api.get_status(tweet_id, tweet_mode='extended')
        tweet_data = json.dumps(tweet._json)
        with open(os.path.join(folder_name, tweet_json_file), mode='a') as file:
            file.write(tweet_data)
            file.write('\n')
        print(str(tweet_id) + 'success')
        current_record += 1
        print(str((current_record / number_of_tweets) * 100) + '% complete')
    except:
        print(str(tweet_id) + 'fail')
        current_record += 1
        print(str((current_record / number_of_tweets) * 100) + '% complete')

892420643555336193success
0.042444821731748725% complete
892177421306343426success
0.08488964346349745% complete
891815181378084864success
0.1273344651952462% complete
891689557279858688success
0.1697792869269949% complete
891327558926688256success
0.21222410865874364% complete
891087950875897856success
0.2546689303904924% complete
890971913173991426success
0.29711375212224106% complete
890729181411237888success
0.3395585738539898% complete
890609185150312448success
0.38200339558573854% complete
890240255349198849success
0.4244482173174873% complete
890006608113172480success
0.46689303904923596% complete
889880896479866881success
0.5093378607809848% complete
889665388333682689success
0.5517826825127334% complete
889638837579907072success
0.5942275042444821% complete
889531135344209921success
0.6366723259762309% complete
889278841981685760success
0.6791171477079796% complete
888917238123831296success
0.7215619694397283% complete
888804989199671297success
0.7640067911714771% complete
888

862831371563274240success
6.451612903225806% complete
862722525377298433success
6.494057724957555% complete
862457590147678208success
6.536502546689304% complete
862096992088072192success
6.578947368421052% complete
861769973181624320fail
6.621392190152801% complete
861383897657036800success
6.66383701188455% complete
861288531465048066success
6.706281833616298% complete
861005113778896900success
6.748726655348047% complete
860981674716409858success
6.791171477079796% complete
860924035999428608success
6.833616298811545% complete
860563773140209665success
6.8760611205432935% complete
860524505164394496success
6.918505942275043% complete
860276583193509888success
6.960950764006792% complete
860184849394610176success
7.00339558573854% complete
860177593139703809success
7.045840407470289% complete
859924526012018688success
7.088285229202038% complete
859851578198683649success
7.130730050933787% complete
859607811541651456success
7.173174872665535% complete
859196978902773760success
7.2156

836397794269200385success
12.903225806451612% complete
836380477523124226success
12.945670628183361% complete
836260088725786625success
12.98811544991511% complete
836001077879255040success
13.03056027164686% complete
835685285446955009success
13.073005093378608% complete
835574547218894849success
13.115449915110355% complete
835536468978302976success
13.157894736842104% complete
835309094223372289success
13.200339558573853% complete
835297930240217089success
13.242784380305602% complete
835264098648616962success
13.285229202037351% complete
835246439529840640success
13.3276740237691% complete
835172783151792128success
13.37011884550085% complete
835152434251116546success
13.412563667232597% complete
834931633769889797success
13.455008488964346% complete
834786237630337024success
13.497453310696095% complete
834574053763584002success
13.539898132427844% complete
834477809192075265success
13.582342954159593% complete
834458053273591808success
13.624787775891342% complete
834209720923721

818588835076603904success
19.269949066213922% complete
818536468981415936success
19.31239388794567% complete
818307523543449600success
19.35483870967742% complete
818259473185828864success
19.39728353140917% complete
818145370475810820success
19.439728353140918% complete
817908911860748288success
19.482173174872667% complete
817827839487737858success
19.524617996604416% complete
817777686764523521success
19.56706281833616% complete
817536400337801217success
19.60950764006791% complete
817502432452313088success
19.65195246179966% complete
817423860136083457success
19.69439728353141% complete
817415592588222464success
19.736842105263158% complete
817181837579653120success
19.779286926994907% complete
817171292965273600success
19.821731748726656% complete
817120970343411712success
19.864176570458405% complete
817056546584727552success
19.906621392190154% complete
816829038950027264success
19.949066213921903% complete
816816676327063552success
19.99151103565365% complete
816697700272001025

798585098161549313success
25.67911714770798% complete
798576900688019456success
25.721561969439726% complete
798340744599797760success
25.76400679117148% complete
798209839306514432success
25.806451612903224% complete
797971864723324932success
25.848896434634977% complete
797545162159308800success
25.891341256366722% complete
797236660651966464success
25.93378607809847% complete
797165961484890113success
25.97623089983022% complete
796904159865868288success
26.01867572156197% complete
796865951799083009success
26.06112054329372% complete
796759840936919040success
26.103565365025467% complete
796563435802726400success
26.146010186757216% complete
796484825502875648success
26.188455008488965% complete
796387464403357696success
26.23089983022071% complete
796177847564038144success
26.273344651952463% complete
796149749086875649success
26.31578947368421% complete
796125600683540480success
26.35823429541596% complete
796116448414461957success
26.400679117147707% complete
796080075804475393s

778748913645780993success
32.088285229202036% complete
778650543019483137success
32.130730050933785% complete
778624900596654080success
32.173174872665534% complete
778408200802557953success
32.21561969439728% complete
778396591732486144success
32.25806451612903% complete
778383385161035776success
32.30050933786078% complete
778286810187399168success
32.34295415959253% complete
778039087836069888success
32.38539898132428% complete
778027034220126208success
32.42784380305603% complete
777953400541634568success
32.47028862478778% complete
777885040357281792success
32.512733446519526% complete
777684233540206592success
32.555178268251275% complete
777641927919427584success
32.597623089983024% complete
777621514455814149success
32.64006791171477% complete
777189768882946048success
32.682512733446515% complete
776819012571455488success
32.72495755517827% complete
776813020089548800success
32.76740237691001% complete
776477788987613185success
32.80984719864177% complete
776249906839351296suc

757729163776290825fail
38.58234295415959% complete
757725642876129280fail
38.62478777589134% complete
757611664640446465fail
38.66723259762309% complete
757597904299253760fail
38.70967741935484% complete
757596066325864448fail
38.75212224108659% complete
757400162377592832fail
38.79456706281834% complete
757393109802180609fail
38.83701188455008% complete
757354760399941633fail
38.879456706281836% complete
756998049151549440fail
38.92190152801358% complete
756939218950160384fail
38.964346349745334% complete
756651752796094464fail
39.006791171477076% complete
756526248105566208fail
39.04923599320883% complete
756303284449767430fail
39.091680814940574% complete
756288534030475264fail
39.13412563667232% complete
756275833623502848fail
39.17657045840407% complete
755955933503782912fail
39.21901528013582% complete
755206590534418437fail
39.26146010186757% complete
755110668769038337fail
39.30390492359932% complete
754874841593970688fail
39.346349745331075% complete
754856583969079297fail
39.

740359016048689152fail
45.458404074702884% complete
740214038584557568fail
45.50084889643464% complete
739979191639244800fail
45.54329371816638% complete
739932936087216128fail
45.58573853989814% complete
739844404073074688fail
45.62818336162988% complete
739623569819336705fail
45.67062818336163% complete
739606147276148736fail
45.71307300509338% complete
739544079319588864fail
45.75551782682513% complete
739485634323156992fail
45.797962648556876% complete
739238157791694849fail
45.840407470288625% complete
738891149612572673fail
45.882852292020374% complete
738885046782832640fail
45.92529711375212% complete
738883359779196928fail
45.96774193548387% complete
738537504001953792fail
46.01018675721562% complete
738402415918125056fail
46.05263157894737% complete
738184450748633089fail
46.09507640067912% complete
738166403467907072fail
46.13752122241087% complete
738156290900254721fail
46.17996604414261% complete
737826014890496000fail
46.222410865874366% complete
737800304142471168fail
46.

713175907180089344fail
52.33446519524618% complete
712809025985978368fail
52.37691001697793% complete
712717840512598017fail
52.41935483870967% complete
712668654853337088fail
52.46179966044142% complete
712438159032893441fail
52.50424448217318% complete
712309440758808576fail
52.54668930390493% complete
712097430750289920fail
52.58913412563667% complete
712092745624633345fail
52.63157894736842% complete
712085617388212225fail
52.67402376910017% complete
712065007010385924fail
52.71646859083192% complete
711998809858043904fail
52.75891341256367% complete
711968124745228288fail
52.801358234295414% complete
711743778164514816fail
52.84380305602716% complete
711732680602345472fail
52.88624787775892% complete
711694788429553666fail
52.92869269949067% complete
711652651650457602fail
52.97113752122241% complete
711363825979756544fail
53.01358234295416% complete
711306686208872448fail
53.0560271646859% complete
711008018775851008fail
53.098471986417664% complete
710997087345876993fail
53.1409

700002074055016451fail
59.210526315789465% complete
699801817392291840fail
59.25297113752123% complete
699788877217865730fail
59.29541595925297% complete
699779630832685056fail
59.33786078098472% complete
699775878809702401fail
59.38030560271647% complete
699691744225525762fail
59.42275042444821% complete
699446877801091073fail
59.46519524617997% complete
699434518667751424fail
59.507640067911716% complete
699423671849451520fail
59.550084889643465% complete
699413908797464576fail
59.59252971137521% complete
699370870310113280fail
59.63497453310696% complete
699323444782047232fail
59.67741935483871% complete
699088579889332224fail
59.71986417657046% complete
699079609774645248fail
59.7623089983022% complete
699072405256409088fail
59.80475382003395% complete
699060279947165696fail
59.84719864176571% complete
699036661657767936fail
59.88964346349746% complete
698989035503689728fail
59.9320882852292% complete
698953797952008193fail
59.97453310696095% complete
698907974262222848fail
60.0169

688828561667567616fail
66.08658743633276% complete
688804835492233216fail
66.12903225806451% complete
688789766343622656fail
66.17147707979628% complete
688547210804498433fail
66.21392190152801% complete
688519176466644993fail
66.25636672325976% complete
688385280030670848fail
66.29881154499151% complete
688211956440801280fail
66.34125636672326% complete
688179443353796608fail
66.383701188455% complete
688116655151435777fail
66.42614601018676% complete
688064179421470721fail
66.4685908319185% complete
687841446767013888fail
66.51103565365025% complete
687826841265172480fail
66.553480475382% complete
687818504314159109fail
66.59592529711375% complete
687807801670897665fail
66.6383701188455% complete
687732144991551489fail
66.68081494057725% complete
687704180304273409fail
66.723259762309% complete
687664829264453632fail
66.76570458404075% complete
687494652870668288fail
66.8081494057725% complete
687480748861947905fail
66.85059422750425% complete
687476254459715584fail
66.893039049236% 

680176173301628928fail
72.96264855687606% complete
680161097740095489fail
73.00509337860781% complete
680145970311643136fail
73.04753820033956% complete
680130881361686529fail
73.08998302207131% complete
680115823365742593fail
73.13242784380306% complete
680100725817409536fail
73.1748726655348% complete
680085611152338944fail
73.21731748726656% complete
680070545539371008fail
73.2597623089983% complete
680055455951884288fail
73.30220713073004% complete
679877062409191424fail
73.3446519524618% complete
679872969355714560fail
73.38709677419355% complete
679862121895714818fail
73.4295415959253% complete
679854723806179328fail
73.47198641765704% complete
679844490799091713fail
73.5144312393888% complete
679828447187857408fail
73.55687606112055% complete
679777920601223168fail
73.5993208828523% complete
679736210798047232fail
73.64176570458405% complete
679729593985699840fail
73.68421052631578% complete
679722016581222400fail
73.72665534804754% complete
679530280114372609fail
73.76910016977

675006312288268288fail
79.83870967741935% complete
675003128568291329fail
79.88115449915111% complete
674999807681908736fail
79.92359932088286% complete
674805413498527744fail
79.9660441426146% complete
674800520222154752fail
80.00848896434634% complete
674793399141146624fail
80.05093378607809% complete
674790488185167872fail
80.09337860780985% complete
674788554665512960fail
80.1358234295416% complete
674781762103414784fail
80.17826825127334% complete
674774481756377088fail
80.22071307300509% complete
674767892831932416fail
80.26315789473685% complete
674764817387900928fail
80.3056027164686% complete
674754018082705410fail
80.34804753820033% complete
674752233200820224fail
80.39049235993208% complete
674743008475090944fail
80.43293718166383% complete
674742531037511680fail
80.4753820033956% complete
674739953134403584fail
80.51782682512734% complete
674737130913071104fail
80.56027164685908% complete
674690135443775488fail
80.60271646859083% complete
674670581682434048fail
80.645161290

671538301157904385fail
86.71477079796264% complete
671536543010570240fail
86.75721561969439% complete
671533943490011136fail
86.79966044142614% complete
671528761649688577fail
86.8421052631579% complete
671520732782923777fail
86.88455008488964% complete
671518598289059840fail
86.92699490662139% complete
671511350426865664fail
86.96943972835314% complete
671504605491109889fail
87.0118845500849% complete
671497587707535361fail
87.05432937181664% complete
671488513339211776fail
87.09677419354838% complete
671486386088865792fail
87.13921901528013% complete
671485057807351808fail
87.18166383701188% complete
671390180817915904fail
87.22410865874363% complete
671362598324076544fail
87.26655348047538% complete
671357843010908160fail
87.30899830220713% complete
671355857343524864fail
87.35144312393888% complete
671347597085433856fail
87.39388794567063% complete
671186162933985280fail
87.43633276740238% complete
671182547775299584fail
87.47877758913413% complete
671166507850801152fail
87.5212224

668636665813057536fail
93.59083191850594% complete
668633411083464705fail
93.63327674023769% complete
668631377374486528fail
93.67572156196944% complete
668627278264475648fail
93.71816638370119% complete
668625577880875008fail
93.76061120543294% complete
668623201287675904fail
93.80305602716469% complete
668620235289837568fail
93.84550084889644% complete
668614819948453888fail
93.88794567062818% complete
668587383441514497fail
93.93039049235993% complete
668567822092664832fail
93.97283531409168% complete
668544745690562560fail
94.01528013582343% complete
668542336805281792fail
94.05772495755518% complete
668537837512433665fail
94.10016977928693% complete
668528771708952576fail
94.14261460101866% complete
668507509523615744fail
94.18505942275043% complete
668496999348633600fail
94.22750424448218% complete
668484198282485761fail
94.26994906621393% complete
668480044826800133fail
94.31239388794566% complete
668466899341221888fail
94.35483870967742% complete
668297328638447616fail
94.39728

In [7]:
# Reading the txt file created above line by line into a list and just pulling out
# the id, retweet_count, and favorite_count

tweet_counts = []

with open('tweet_json.txt', 'r') as file:
    for line in file.readlines():
        this_tweet = json.loads(line)
        this_tweet_data = [this_tweet['id'], this_tweet['retweet_count'], this_tweet['favorite_count']]
        tweet_counts.append(this_tweet_data)

In [8]:
tweet_counts_df = pd.DataFrame(tweet_counts, columns=['tweet_id', 'retweet_count', 'favorite_count'])
tweet_counts_df

tweet_id  retweet_count  favorite_count
0    892420643555336193           7452           35291
1    892177421306343426           5534           30537
2    891815181378084864           3660           22984
3    891689557279858688           7624           38585
4    891327558926688256           8218           36867
..                  ...            ...             ...
872  759159934323924993           1128               0
873  759099523532779520           4054           14242
874  759047813560868866           1956            6364
875  758854675097526272            879            3484
876  758828659922702336           3760           11004

[877 rows x 3 columns]

In [9]:
tweet_archive_master = archive.merge(tweet_counts_df, how='left', on='tweet_id')
tweet_archive_master

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
...                  ...                    ...                  ...   
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
0     2017-08-01 16:23:56 +0000   
1     2017-08-01 00:17:27 +0000   
2     2017-07-31 00:18:03 +0000   
3     2017-07-30 15:58:51 +0000   
4     2017-07-29 16:00:24 +0000   
...                         ...   
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
0     This is Phineas. He's a mystical boy. Only eve...                  NaN   
1     This is Tilly. She's just checking pup on you....                  NaN   
2     This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3     This is Darla. She commenced a snooze mid meal...                  NaN   
4     This is Franklin. He would like you to stop ca...                  NaN   
...                                                 ...                  ...   
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
0                          NaN                        NaN   
1                          NaN                        NaN   
2                          NaN                        NaN   
3                          NaN                        NaN   
4                          NaN                        NaN   
...                        ...                        ...   
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
0     https://twitter.com/dog_rates/status/892420643...                13   
1     https://twitter.com/dog_rates/status/892177421...                13   
2     https://twitter.com/dog_rates/status/891815181...                12   
3     https://twitt

## Assessing the Data ##

In [10]:
tweet_archive_master

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
...                  ...                    ...                  ...   
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
0     2017-08-01 16:23:56 +0000   
1     2017-08-01 00:17:27 +0000   
2     2017-07-31 00:18:03 +0000   
3     2017-07-30 15:58:51 +0000   
4     2017-07-29 16:00:24 +0000   
...                         ...   
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
0     This is Phineas. He's a mystical boy. Only eve...                  NaN   
1     This is Tilly. She's just checking pup on you....                  NaN   
2     This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3     This is Darla. She commenced a snooze mid meal...                  NaN   
4     This is Franklin. He would like you to stop ca...                  NaN   
...                                                 ...                  ...   
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
0                          NaN                        NaN   
1                          NaN                        NaN   
2                          NaN                        NaN   
3                          NaN                        NaN   
4                          NaN                        NaN   
...                        ...                        ...   
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
0     https://twitter.com/dog_rates/status/892420643...                13   
1     https://twitter.com/dog_rates/status/892177421...                13   
2     https://twitter.com/dog_rates/status/891815181...                12   
3     https://twitt

In [11]:
tweet_archive_master.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2356 entries, 0 to 2355
Data columns (total 19 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [12]:
tweet_archive_master[['rating_numerator', 'retweet_count', 'favorite_count']].describe()

rating_numerator  retweet_count  favorite_count
count       2356.000000     877.000000      877.000000
mean          13.126486    4593.530217    13540.438997
std           45.876648    5176.630312    14223.492377
min            0.000000       1.000000        0.000000
25%           10.000000    1914.000000     5109.000000
50%           11.000000    3144.000000    10333.000000
75%           12.000000    5389.000000    19030.000000
max         1776.000000   54847.000000   129746.000000

In [13]:
tweet_archive_master['rating_denominator'].describe()

count    2356.000000
mean       10.455433
std         6.745237
min         0.000000
25%        10.000000
50%        10.000000
75%        10.000000
max       170.000000
Name: rating_denominator, dtype: float64

In [14]:
tweet_archive_master[tweet_archive_master['rating_denominator'] > 10][['tweet_id', 'rating_denominator']]

tweet_id  rating_denominator
342   832088576586297345                  15
433   820690176645140481                  70
784   775096608509886464                  11
902   758467244762497024                 150
1068  740373189193256964                  11
1120  731156023742988288                 170
1165  722974582966214656                  20
1202  716439118184652801                  50
1228  713900603437621249                  90
1254  710658690886586372                  80
1274  709198395643068416                  50
1351  704054845121142784                  50
1433  697463031882764288                  40
1598  686035780142297088                  20
1634  684225744407494656                 130
1635  684222868335505415                 110
1662  682962037429899265                  11
1663  682808988178739200                  16
1779  677716515794329600                 120
1843  675853064436391936                  80

In [15]:
tweet_archive_master[tweet_archive_master['rating_denominator'] > 10][['tweet_id', 'rating_denominator']].count()

tweet_id              20
rating_denominator    20
dtype: int64

In [16]:
tweet_archive_master.name.sample(20)

267       Penny
1150       None
96         None
405        Cash
324      Lipton
1594       Birf
586      Tucker
251        None
2082       Sage
1938       None
318        None
114        Cody
1448       Berb
1241    Chester
616        None
1355    Scooter
1711      Percy
328      Bronte
2086       None
1569    Trooper
Name: name, dtype: object

In [17]:
tweet_archive_master.groupby('name')['tweet_id'].count()

name
Abby            2
Ace             1
Acro            1
Adele           1
Aiden           1
               ..
such            1
the             8
this            1
unacceptable    1
very            5
Name: tweet_id, Length: 957, dtype: int64

In [18]:
tweet_archive_master.groupby('name')['tweet_id'].count().sort_values()

name
Kingsley      1
Julius        1
Sora          1
Kaia          1
Kaiya         1
           ... 
Lucy         11
Oliver       11
Charlie      12
a            55
None        745
Name: tweet_id, Length: 957, dtype: int64

In [21]:
# Running a loop to check how many tweet texts contain mention of the 
# dog stages (pupper, puppo, doggo, floofer) as compared to the information in the columns

def dog_counts(string):
    appearances = 0
    for entry in tweet_archive_master.text:
        if string in entry:
            appearances += 1
    
    number_in_archive = tweet_archive_master.groupby(string)['text'].count()[string]
    print("Difference in text and archive of the appearance of " + string + ":")
    print(number_in_archive - appearances)

In [23]:
dog_counts('pupper')

Difference in text and archive of the appearance of pupper:
-15


In [24]:
dog_counts('doggo')

Difference in text and archive of the appearance of doggo:
-1


In [25]:
dog_counts('floofer')

Difference in text and archive of the appearance of floofer:
6


In [26]:
dog_counts('puppo')

Difference in text and archive of the appearance of puppo:
-7


In [27]:
image_predictions.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [28]:
image_predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [29]:
image_predictions.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

### Dataset Issues ###

- Quality Issues
    1. There are 181 retweets
    2. Only 877 tweets successfully pulled retweet and favorite count data
    3. There are a slew of incorrect ratings
        - some grabbed a date or other number instead of the rating
        - many are ratings based on multiple dogs in the same post
    4. Some of the names are incorrect
        - 55 records list 'a'
        - some records list random words
    5. The appearance of the dog stages does not appear at first glance to be accurately recorded in those columns
    6. 'retweet_count' and 'favorite_count' were imported as floats
    7. The 'expanded_urls' column often lists the same URL multiple times
    8. Timestamp is saved as a string


- Tidiness Issues
    1. Should be one dataset, as the photo information is unique to each tweet
    2. The retweet information columns are unnecessary since none should be retweets
    3. Dog stage is a single variable recorded in four columns, and that data should be categorical

## Cleaning the Data ##

### Quality Issues ###

#### 1. There are 181 retweets ####

**Define:** Remove the lines that are retweets using the rows that have retweeted_status_id as non-null.

**Code:**

In [30]:
retweet_indices = tweet_archive_master[tweet_archive_master.retweeted_status_id.isnull() == False].index
tweet_archive_master = tweet_archive_master.drop(retweet_indices)

**Test:**

In [31]:
tweet_archive_master[tweet_archive_master.retweeted_status_id.isnull() == False]['tweet_id'].count()

0

#### 2. Only 877 tweets successfully pulled retweet and favorite count data ####

Define: Remove rows that have null values in retweet_count and favorite_count.

Code:

In [32]:
missing_count_indices = tweet_archive_master[tweet_archive_master.favorite_count.isnull() == True].index
tweet_archive_master = tweet_archive_master.drop(missing_count_indices)

**Test:**

In [33]:
tweet_archive_master[tweet_archive_master.retweet_count.isnull() == True]['tweet_id'].count()

0

In [34]:
tweet_archive_master[tweet_archive_master.favorite_count.isnull() == True]['tweet_id'].count()

0

#### 3. Incorrect Ratings ####

**Define:** Identify and correct any ratings that do not have 10 as the denominator and are not an integer between 10-14 in the numerator. There are posts that have ratings under 10/10, but they are generally either not dogs or pictures where the account is pretending not to see a dog in the photo. The only 15 in the rating column was the account stating that they were still looking for their first 15/10 rating, so anything over 14 is not valid.

Note that a lot of the individual id issues that were listed were removed when the retweets and tweets with favorite_count and retweet_count were removed.

**Code:**

In [35]:
incorrect_denominators = tweet_archive_master[tweet_archive_master.rating_denominator != 10].index

In [36]:
incorrect_high_numerators = tweet_archive_master[tweet_archive_master.rating_numerator > 14].index

In [37]:
incorrect_low_numerators = tweet_archive_master[tweet_archive_master.rating_numerator < 10].index

In [38]:
incorrect_ratings = incorrect_denominators.append([incorrect_high_numerators, incorrect_low_numerators])

In [39]:
tweet_archive_master = tweet_archive_master.drop(incorrect_ratings)

**Test:**

In [40]:
tweet_archive_master.groupby('rating_denominator').count()['tweet_id']

rating_denominator
10    703
Name: tweet_id, dtype: int64

In [41]:
tweet_archive_master.groupby('rating_numerator').count()['tweet_id']

rating_numerator
10     46
11    132
12    259
13    227
14     39
Name: tweet_id, dtype: int64

#### 4. Some of the names are incorrect ####

**Define:** Verify that the name fields actually show names or are null. I think the only way to do this is to create a separate list of the names, remove duplicates, and then check manually.

**Code:**

In [42]:
list_of_names = tweet_archive_master.groupby('name').count().index
list_of_names

Index(['Aja', 'Akumi', 'Al', 'Albert', 'Albus', 'Alexander', 'Alf', 'Alfie',
       'Alfy', 'Alice',
       ...
       'Zoey', 'Zooey', 'a', 'incredibly', 'my', 'not', 'one', 'quite', 'such',
       'very'],
      dtype='object', name='name', length=417)

In [43]:
with pd.option_context("display.max_rows", 1000):
    for entry in list_of_names:
        print(entry)

Aja
Akumi
Al
Albert
Albus
Alexander
Alf
Alfie
Alfy
Alice
Anakin
Angel
Anna
Archie
Arlo
Arnie
Arya
Ash
Aspen
Atlas
Aubie
Augie
Autumn
Ava
Bailey
Balto
Barney
Baron
Bauer
BeBe
Bear
Beau
Beebop
Bella
Belle
Benedict
Benji
Benny
Bentley
Betty
Binky
Blue
Bo
Bonaparte
Bones
Boomer
Boots
Brady
Brandi
Brat
Brody
Bronte
Brownie
Bruce
Brudge
Bruno
Brutus
Buddy
Bungalo
Burt
Butter
Cali
Callie
Canela
Cannon
Carbon
Carl
Cash
Cassie
Cermet
Charlie
Chef
Chelsea
Chester
Chevy
Chip
Chipson
Chloe
Chubbs
Clark
Clifford
Coco
Cody
Combo
Comet
Cooper
Craig
Crawford
Cupid
Daisy
Dakota
Dale
Dallas
Dante
Darla
Dash
Dave
Davey
Dawn
Deacon
Derek
Dewey
Dexter
Dido
Diogi
Dixie
Django
Dobby
Doc
DonDon
Doobert
Duchess
Duddles
Dudley
Duke
Dutch
Earl
Eevee
Einstein
Eleanor
Eli
Elliot
Emanuel
Emma
Emmy
Eugene
Finley
Finn
Fiona
Fizz
Flash
Florence
Frank
Frankie
Franklin
Furzey
Gabe
Gary
General
George
Georgie
Gerald
Gidget
Ginger
Gizmo
Glenn
Godi
Goose
Grey
Grizzwald
Gromit
Gus
Halo
Hank
Happy
Harlso
Harold
Heinrich
Herc

In [44]:
problem_names = ['O', 'a', 'incredibly', 'my', 'not', 'one', 'quite', 'such', 'very']

In [45]:
problem_name_df = pd.DataFrame()
problem_name_df['text'] = tweet_archive_master[tweet_archive_master.name.isin(problem_names)]['text']
problem_name_df['name'] = tweet_archive_master[tweet_archive_master.name.isin(problem_names)]['name']
problem_name_df

text        name
22   I've yet to rate a Venezuelan Hover Wiener. Th...        such
56   Here is a pupper approaching maximum borkdrive...           a
169  We only rate dogs. This is quite clearly a smo...       quite
193  Guys, we only rate dogs. This is quite clearly...       quite
335  There's going to be a dog terminal at JFK Airp...         not
369  Occasionally, we're sent fantastic stories. Th...         one
542  We only rate dogs. Please stop sending in non-...  incredibly
649  Here is a perfect example of someone who has t...           a
775  This is O'Malley. That is how he sleeps. Doesn...           O
801  Guys this is getting so out of hand. We only r...           a
819  We only rate dogs. Pls stop sending in non-can...        very
852  This is my dog. Her name is Zoey. She knows I'...          my

In [46]:
problem_name_df['correct_name'] = ['Roxy', 'Puffie', None, None, None, 'Grace', None, 'Forrest', "O'Malley", None, None, 'Zoey']
problem_name_indices = problem_name_df.index

In [47]:
for row in problem_name_df.index:
    updated_name = problem_name_df.at[row, 'correct_name']
    tweet_archive_master.at[row, 'name'] = updated_name

**Test:**

In [48]:
tweet_archive_master[tweet_archive_master.index.isin(problem_name_indices)]['name']

22         Roxy
56       Puffie
169        None
193        None
335        None
369       Grace
542        None
649     Forrest
775    O'Malley
801        None
819        None
852        Zoey
Name: name, dtype: object

#### 5. The appearance of the dog stages is not accurately recorded ####

**Define:** Run a loop to pull out rows that have 'pupper', 'puppo', 'doggo', and 'floofer' in the text but do not list it in that column and vice versa, and then correct the value.

**Code:**

In [49]:
mismatched_dog_stage = []
def dog_stage_loops(string):
        if (string in row.text.lower() and row[string] != string) or (string not in row.text.lower() and row[string] == string):
                mismatched_dog_stage.append(row)

for i, row in tweet_archive_master.iterrows():
    dog_stage_loops('doggo')
    dog_stage_loops('floofer')
    dog_stage_loops('pupper')
    dog_stage_loops('puppo')

mismatched_dog_stage = pd.DataFrame(mismatched_dog_stage)
mismatched_dog_stage

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
54   881666595344535552                    NaN                  NaN   
83   876537666061221889                    NaN                  NaN   
85   876120275196170240                    NaN                  NaN   
106  871879754684805121                    NaN                  NaN   
134  866686824827068416                    NaN                  NaN   
172  858843525470990336                    NaN                  NaN   
228  848213670039564288           8.482121e+17         4.196984e+09   
268  841439858740625411                    NaN                  NaN   
274  840698636975636481           8.406983e+17         8.405479e+17   
477  815990720817401858                    NaN                  NaN   
536  807010152071229440                    NaN                  NaN   
545  805826884734976000                    NaN                  NaN   
798  772877495989305348                    NaN                  NaN   

                     timestamp  \
54   2017-07-03 00:11:11 +0000   
83   2017-06-18 20:30:39 +0000   
85   2017-06-17 16:52:05 +0000   
106  2017-06-06 00:01:46 +0000   
134  2017-05-22 16:06:55 +0000   
172  2017-05-01 00:40:27 +0000   
228  2017-04-01 16:41:12 +0000   
268  2017-03-14 00:04:30 +0000   
274  2017-03-11 22:59:09 +0000   
477  2017-01-02 18:38:42 +0000   
536  2016-12-08 23:53:08 +0000   
545  2016-12-05 17:31:15 +0000   
798  2016-09-05 19:22:09 +0000   

                                                source  \
54   <a href="http://twitter.com/download/iphone" r...   
83   <a href="http://twitter.com/download/iphone" r...   
85   <a href="http://twitter.com/download/iphone" r...   
106  <a href="http://twitter.com/download/iphone" r...   
134  <a href="http://twitter.com/download/iphone" r...   
172  <a href="http://twitter.com/download/iphone" r...   
228  <a href="http://twitter.com/download/iphone" r...   
268  <a href="http://twitter.com/download/iphone" r...   
274  <a href="http://twitter.com/download/iphone" r...   
477  <a href="http://twitter.com/download/iphone" r...   
536  <a href="http://twitter.com/download/iphone" r...   
545  <a href="http://twitter.com/download/iphone" r...   
798  <a href="http://twitter.com" rel="nofollow">Tw...   

                                                  text  retweeted_status_id  \
54   This is Gary. He couldn't miss this puppertuni...                  NaN   
83   I can say with the pupmost confidence that the...                  NaN   
85   Meet Venti, a seemingly caffeinated puppoccino...                  NaN   
106  Say hello to Lassie. She's celebrating #PrideM...                  NaN   
134  This is Lili. She can't believe you betrayed h...                  NaN   
172  I have stumbled puppon a doggo painting party....                  NaN   
228  Jerry just apuppologized to me. He said there ...                  NaN   
268  Here we have some incredible doggos for #K9Vet...                  NaN   
274  @0_kelvin_0 &gt;10/10 is reserved for puppos s...                  NaN   
477  Meet Jack. He's one of the rare doggos that do...                  NaN   
536  This is Lennon. He's a Boopershnoop Pupperdoop...                  NaN   
545  This is Duke. He is not a fan of the pupporazz...                  NaN   
798  You need to watch these two doggos argue throu...                  NaN   

     retweeted_status_user_id  retweeted_status_timestamp  \
54                        NaN                         NaN   
83                        NaN                         NaN   
85                        NaN                         NaN   
106                       NaN                         NaN   
134                       NaN                         NaN   
172                       NaN                         NaN   
228                       NaN                         NaN   
268                       NaN                         NaN   
274                       NaN                         NaN   
477          

Once I ran these loops, the values that appeared were all situations where the string appeared in the text because of a pun. The only one that seemed to truly not correspond was a record referred to with 'doggo', where it was actually multiple dogs, index 172.

In [50]:
tweet_archive_master.loc[172]

tweet_id                                                     858843525470990336
in_reply_to_status_id                                                       NaN
in_reply_to_user_id                                                         NaN
timestamp                                             2017-05-01 00:40:27 +0000
source                        <a href="http://twitter.com/download/iphone" r...
text                          I have stumbled puppon a doggo painting party....
retweeted_status_id                                                         NaN
retweeted_status_user_id                                                    NaN
retweeted_status_timestamp                                                  NaN
expanded_urls                 https://twitter.com/dog_rates/status/858843525...
rating_numerator                                                             13
rating_denominator                                                           10
name                                    

In [51]:
tweet_archive_master.at[172, 'doggo'] = None

#### 6. 'retweet_count' and 'favorite_count' were imported as floats ####

**Define:** Convert those column data types to integers.

**Code:**

In [52]:
tweet_archive_master['retweet_count'] = tweet_archive_master['retweet_count'].astype(int)
tweet_archive_master['favorite_count'] = tweet_archive_master['favorite_count'].astype(int)

**Test:**

In [53]:
tweet_archive_master.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 703 entries, 0 to 899
Data columns (total 19 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    703 non-null    int64  
 1   in_reply_to_status_id       25 non-null     float64
 2   in_reply_to_user_id         25 non-null     float64
 3   timestamp                   703 non-null    object 
 4   source                      703 non-null    object 
 5   text                        703 non-null    object 
 6   retweeted_status_id         0 non-null      float64
 7   retweeted_status_user_id    0 non-null      float64
 8   retweeted_status_timestamp  0 non-null      object 
 9   expanded_urls               680 non-null    object 
 10  rating_numerator            703 non-null    int64  
 11  rating_denominator          703 non-null    int64  
 12  name                        697 non-null    object 
 13  doggo                       702 non

#### 7. The 'expanded_urls' column often lists the same URL multiple times ####

**Define:** Separate the string in the column by commas, and if the subsequent strings are duplicates, delete them, and then resave the corrected string into the 'expanded_urls' column.

**Code:**

In [54]:
tweet_archive_master.loc[898]['expanded_urls']

'https://twitter.com/dog_rates/status/758854675097526272/photo/1,https://twitter.com/dog_rates/status/758854675097526272/photo/1,https://twitter.com/dog_rates/status/758854675097526272/photo/1,https://twitter.com/dog_rates/status/758854675097526272/photo/1'

In [55]:
for i, row in tweet_archive_master.iterrows():
    url_string = row.expanded_urls
    if type(url_string) == float:
        # NaN values broke the loop since the functions below do not work on floats
        pass
    else:
        string_list = url_string.split(",")
        string_list = list(set(string_list))
        tweet_archive_master.at[i, 'expanded_urls'] = ','.join(string_list)

**Test:**

In [56]:
tweet_archive_master.loc[898]['expanded_urls']

'https://twitter.com/dog_rates/status/758854675097526272/photo/1'

#### 8. Timestamp is saved as a string ####

**Define:** Convert the timestamp column to a date/time format.

**Code:**

In [57]:
tweet_archive_master.timestamp

0      2017-08-01 16:23:56 +0000
1      2017-08-01 00:17:27 +0000
2      2017-07-31 00:18:03 +0000
3      2017-07-30 15:58:51 +0000
4      2017-07-29 16:00:24 +0000
                 ...            
893    2016-07-30 17:51:13 +0000
894    2016-07-30 01:22:17 +0000
897    2016-07-29 15:27:55 +0000
898    2016-07-29 02:40:28 +0000
899    2016-07-29 00:57:05 +0000
Name: timestamp, Length: 703, dtype: object

In [58]:
tweet_archive_master.timestamp = pd.to_datetime(tweet_archive_master.timestamp, format='%Y/%m/%d %H:%M:%S')

**Test:**

In [59]:
tweet_archive_master.timestamp

0     2017-08-01 16:23:56+00:00
1     2017-08-01 00:17:27+00:00
2     2017-07-31 00:18:03+00:00
3     2017-07-30 15:58:51+00:00
4     2017-07-29 16:00:24+00:00
                 ...           
893   2016-07-30 17:51:13+00:00
894   2016-07-30 01:22:17+00:00
897   2016-07-29 15:27:55+00:00
898   2016-07-29 02:40:28+00:00
899   2016-07-29 00:57:05+00:00
Name: timestamp, Length: 703, dtype: datetime64[ns, UTC]

### Tidiness Issues ###

#### 1. 'tweet_archive_master' and 'image_predictions' should be one dataset

**Define:** The two datasets should be combined into one, as they each refer to a single variable's information in both. Combining them also allows the possibility of doing analyses based on the variables present in both, such as predicted dog breeds and retweet and favorite counts.

**Code:**

In [60]:
tweet_archive_master = tweet_archive_master.merge(image_predictions, how='left', on='tweet_id')

**Test:**

In [61]:
tweet_archive_master.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 703 entries, 0 to 702
Data columns (total 30 columns):
 #   Column                      Non-Null Count  Dtype              
---  ------                      --------------  -----              
 0   tweet_id                    703 non-null    int64              
 1   in_reply_to_status_id       25 non-null     float64            
 2   in_reply_to_user_id         25 non-null     float64            
 3   timestamp                   703 non-null    datetime64[ns, UTC]
 4   source                      703 non-null    object             
 5   text                        703 non-null    object             
 6   retweeted_status_id         0 non-null      float64            
 7   retweeted_status_user_id    0 non-null      float64            
 8   retweeted_status_timestamp  0 non-null      object             
 9   expanded_urls               680 non-null    object             
 10  rating_numerator            703 non-null    int64             

#### 2. The retweet information columns are unnecessary since none should be retweets ####

**Define:** The retweet information columns should be removed since all retweets were removed from the dataframe.

**Code:**

In [62]:
retweet_columns = ['retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp']

tweet_archive_master.drop(columns=retweet_columns, inplace=True)

**Test:**

In [63]:
tweet_archive_master.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 703 entries, 0 to 702
Data columns (total 27 columns):
 #   Column                 Non-Null Count  Dtype              
---  ------                 --------------  -----              
 0   tweet_id               703 non-null    int64              
 1   in_reply_to_status_id  25 non-null     float64            
 2   in_reply_to_user_id    25 non-null     float64            
 3   timestamp              703 non-null    datetime64[ns, UTC]
 4   source                 703 non-null    object             
 5   text                   703 non-null    object             
 6   expanded_urls          680 non-null    object             
 7   rating_numerator       703 non-null    int64              
 8   rating_denominator     703 non-null    int64              
 9   name                   697 non-null    object             
 10  doggo                  702 non-null    object             
 11  floofer                703 non-null    object             

#### 3. Dog stage is a single variable recorded in four columns ####

**Define:** Dog stage is a single variable and therefore should be in a single, categorical column instead of spread across four columns.

**Code:**

In [64]:
dog_stage_columns = ['doggo', 'floofer', 'pupper', 'puppo']

In [65]:
dog_stages_df = pd.melt(tweet_archive_master, id_vars='tweet_id', value_vars=dog_stage_columns, var_name='dog_stage', 
        value_name='dog_stage_true', ignore_index=False)
dog_stages_df

tweet_id dog_stage dog_stage_true
0    892420643555336193     doggo           None
1    892177421306343426     doggo           None
2    891815181378084864     doggo           None
3    891689557279858688     doggo           None
4    891327558926688256     doggo           None
..                  ...       ...            ...
698  759446261539934208     puppo           None
699  759197388317847553     puppo           None
700  759047813560868866     puppo           None
701  758854675097526272     puppo           None
702  758828659922702336     puppo           None

[2812 rows x 3 columns]

In [66]:
dog_stages_df = dog_stages_df[dog_stages_df.dog_stage_true != 'None']

In [67]:
# Dropping the rows where there are multiple values of the four types, this is a quality
# issue that I did not address in the previous cleaning step but would in a full data cleaning process

dog_stages_df.drop_duplicates(subset='tweet_id', keep='first', inplace=True)

<ipython-input-67-b2b4b6ce1d18>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dog_stages_df.drop_duplicates(subset='tweet_id', keep='first', inplace=True)


In [68]:
tweet_archive_master = tweet_archive_master.merge(dog_stages_df, how='left', on='tweet_id')

In [69]:
tweet_archive_master.drop(columns=dog_stage_columns, inplace=True)
tweet_archive_master.drop(columns='dog_stage_true', inplace=True)

In [70]:
tweet_archive_master.dog_stage = tweet_archive_master.dog_stage.astype('category')

**Test:**

In [71]:
tweet_archive_master.groupby('dog_stage')['tweet_id'].count()

dog_stage
doggo      60
floofer     3
pupper     39
puppo      18
Name: tweet_id, dtype: int64

In [72]:
tweet_archive_master.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 703 entries, 0 to 702
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype              
---  ------                 --------------  -----              
 0   tweet_id               703 non-null    int64              
 1   in_reply_to_status_id  25 non-null     float64            
 2   in_reply_to_user_id    25 non-null     float64            
 3   timestamp              703 non-null    datetime64[ns, UTC]
 4   source                 703 non-null    object             
 5   text                   703 non-null    object             
 6   expanded_urls          680 non-null    object             
 7   rating_numerator       703 non-null    int64              
 8   rating_denominator     703 non-null    int64              
 9   name                   697 non-null    object             
 10  retweet_count          703 non-null    int32              
 11  favorite_count         703 non-null    int32              

In [73]:
# Storing the dataframe as a csv

tweet_archive_master.to_csv('twitter_archive_master.csv', index=False)

#### Sites I used: ####

Converting JSON dictionary object to string:
https://www.geeksforgeeks.org/python-convert-dictionary-object-into-string/

Reading text file line by line:
https://tutorial.eyehunts.com/python/python-read-file-line-by-line-readlines/

Converting JSON strings into objects:
https://www.freecodecamp.org/news/python-read-json-file-how-to-load-json-from-a-file-and-parse-dumps/

Viewing an untruncated column in Jupyter (for double checking the full 'name' column): https://stackoverflow.com/questions/43909817/how-to-stop-jupyter-outputting-truncated-results-when-using-pd-series-value-count

.isin() function for pulling rows out of a DataFrame using a list:
https://stackoverflow.com/questions/19960077/how-to-filter-pandas-dataframe-using-in-and-not-in-like-in-sql

.iterrows() to iterate through each row to double check dog stages:
https://www.stackabuse.com/how-to-iterate-over-rows-in-a-pandas-dataframe/

Removing duplicates from a Python list:
https://www.geeksforgeeks.org/python-ways-to-remove-duplicates-from-list/